### ISSUE TITLE
notifications MVP: mentions

### ISSUE THREAD

**Issue by gyst** [ploneintranet/ploneintranet#371]
As a user,
When another user @mentions me in a post,
I'd like to get notified of that,
So that I can respond and sustain meaningful conversations with my colleagues.

As a user,
When I mention another user in a post,
I expect them to be notified of that,
So that I can use mentions instead of having to send an email.

The notifications subsystem is potentially a big system. A full fledged event listening infrastructure needed to realize e.g. #360 (show "published" events) raises all kinds of architectural questions (like async for "exploding" events and how to summarize truckloads of similar notifications). @mattss rightly suggested we need a sprint to formulate and execute a coherent plan for that.

This ticket is of a much smaller scope. Instead of anticipating the full event listening system described above, this is much simpler: if a user posts a statusupdate that mentions one or more others, use the notification box to show those others that they were mentioned.

A minimal implementation could work as follows:
- mentions are already stored on StatusUpdate objects on creation (verify)
- on handling a statusupdate post (in the statuscontainer.add() method), for each mention, create a corresponding notification for the mentioned user, and store this notification
- on loading a page, the notification box shows the unread notifications count
- on opening the notification box, it shows a list of notifications and automatically marks all notifications as read

NB:
- there's no need for an "all notifications" view at this stage
- a notification of a mention links directly to the relevant post
- there's no need for websocket push to update notifications and unread counter, a page refresh is fine

The upside of this is that this builds the initial notifications roundtrip including backend/frontent integration. We can then later embellish that with async handlers, websockets etc etc.

**Comment by mattss** [comment]
Work in progress on https://github.com/ploneintranet/ploneintranet/tree/basic-notifications

**Comment by mattss** [comment]
@gyst Is there a way to view individual status updates? How do I link from the notifications listing to the status update the notification is for?

**Comment by gyst** [comment]
On 10/07/15 13:47, Matthew Sital-Singh wrote:

> @gyst https://github.com/gyst Is there a way to view individual status
> updates? How do I link from the notifications listing to the status
> update the notification is for?

There is a StatusUpdateView in activitystream but it's a helper view
that IIRC is not directly accessible on a URL yet.

https://github.com/ploneintranet/ploneintranet/blob/master/src/ploneintranet/activitystream/browser/statusupdate.py

It should not be that hard to subclass that into an actual traversable
view on the navigation root that renders an individual statusupdate, or
rather the whole conversation that statusupdate is a part of.

That could be e.g. /portal_url/@@statusconversation/12345 with the
trailing part being the statusid.

For an example of how to extract the statusid from the url that way see
https://github.com/ploneintranet/ploneintranet/blob/master/src/ploneintranet/network/browser/likes.py#L112

Main thing such a view needs to check is that security is not
compromised. I've protected the backend .get() method for fetching a
statusupdate so that should be OK, but it doesn't hurt to double check.

## :*CU#

```
Guido Stevens  |  +31.43.3618933  |  http://cosent.nl

s o c i a l   k n o w l e d g e   t e c h n o l o g y
```

**Comment by mattss** [comment]
Chatted to @pilz about this and it also seems to be possible to link to a status update using an anchor on the dashboard or stream, as each status update has an anchor tag with its unique id.

We still appear to have the @@stream view from plonesocial, although it is not styled. I could either link to that or the dashboard for now, with the assumption that the notification is recent enough to still appear there? I am slightly concerned that adding a specific view for status updates is the start of something bigger and will explode this 'mvp' task!

**Comment by gyst** [comment]
On 10/07/15 15:41, Matthew Sital-Singh wrote:

> Chatted to @pilz https://github.com/pilz about this and it also seems
> to be possible to link to a status update using an anchor on the
> dashboard or stream, as each status update has an anchor tag with its
> unique id.

Not valid. The update may have scrolled out of the page i.e. not be
available.

> We still appear to have the @@stream https://github.com/stream view
> from plonesocial, although it is not styled. I could either link to that
> or the dashboard for now, with the assumption that the notification is
> recent enough to still appear there? I am slightly concerned that adding
> a specific view for status updates is the start of something bigger and
> will explode this 'mvp' task!

I get that. There needs to be a reply widget etc. However I did a lot of
work on these views and am confident this can be derisked - we have a
nice update rendering already inline, it's just a matter of wrapping
that in a full page skeleton. Am happy to work with you on that.

## :*CU#

```
Guido Stevens  |  +31.43.3618933  |  http://cosent.nl

s o c i a l   k n o w l e d g e   t e c h n o l o g y
```

**Comment by gyst** [comment]
Separate view is scheduled (but on hold) in #486. In this ticket I'm OK with linking to the dashboard (portal root) for now. Please don't link to the current page since the notification may be viewed from another URL.

**Comment by mattss** [comment]
Resolved by #483
